source/grid/manifold_lib.cc

```diff
@@ -281,71 +357,149 @@ get_new_point (const ArrayView<const Point<spacedim>> &vertices,
+  double rho = 0.;
+  Tensor<1,spacedim> candidate;
 
-  return center+rho*candidate;
-}
+  // Step 1:
+  // Perform a simple average ...
+  {
+    double total_weights = 0.;
+    for (unsigned int i = 0; i < n_points; i++)
+      {
+        if (std::abs(1-weights[i])<tolerance)
+          return vertices[i];
+        const Tensor<1, spacedim> direction(vertices[i] - center);
+        rho += direction.norm() * weights[i];
+        candidate += direction * weights[i];
+        total_weights += weights[i];
+      }
+    // ... and normalize if the candidate is different from the origin.
+    const double norm = candidate.norm();
+    if (norm == 0.)
+      return center;
+    candidate /= candidate.norm();
+    rho /= total_weights;
+  }
 
-// ============================================================
-// CylindricalManifold
-// ============================================================
+  if (spacedim<2)
+    return center + rho*candidate;
 
-namespace internal
-{
-  namespace CylindricalManifold
+  // Step 2:
+  // Do Newton-style iterations to improve the estimate.
+  //
+  // In this step, we consider all points and directions to be embedded
+  // in a three-dimensional space.
   {
-    namespace
-    {
-      // helper function to compute a vector orthogonal to a given one.
-      template <int spacedim>
-      Point<spacedim>
-      compute_normal(const Tensor<1,spacedim> &vector)
+    Tensor<1, 3> xVec;
+    for (unsigned int c=0; c<spacedim; ++c)
+      xVec[c]=candidate[c];
+
+    // If the candidate happens to coincide with a normalized
+    // direction, we return it. Otherwise, the Hessian would be singular.
+    boost::container::small_vector<Tensor<1, 3>, 100> directions(n_points);
+    for (unsigned int i=0; i<n_points; ++i)
       {
-        AssertThrow(vector.norm() != 0.,
-                    ExcMessage("The direction parameter must not be zero!"));
-        Point<3> normal;
-        if (std::abs(vector[0]) >= std::abs(vector[1])
-            && std::abs(vector[0]) >= std::abs(vector[2]))
-          {
-            normal[1]=-1.;
-            normal[2]=-1.;
-            normal[0]=(vector[1]+vector[2])/vector[0];
-          }
-        else if (std::abs(vector[1]) >= std::abs(vector[0])
-                 && std::abs(vector[1]) >= std::abs(vector[2]))
-          {
-            normal[0]=-1.;
-            normal[2]=-1.;
-            normal[1]=(vector[0]+vector[2])/vector[1];
-          }
-        else
-          {
-            normal[0]=-1.;
-            normal[1]=-1.;
-            normal[2]=(vector[0]+vector[1])/vector[2];
-          }
-        return normal;
+        for (unsigned int c = 0; c < spacedim; ++c)
+          directions[i][c] = vertices[i][c] - center[c];
+        const double norm = directions[i].norm();
+        Assert(norm != 0.,
+               ExcMessage("One of the vertices coincides with the center. "
+                          "This is not allowed!"));
+        directions[i] /= directions[i].norm();
```

Comment:
Can you use `directions[i] /= norm`? (This avoids the warning about the unused variable and was likely intended anyway.)